llvm/lib/Transforms/Utils/LoopIterCountsToAssumptions.cpp

```diff
@@ -0,0 +1,215 @@
+                    << CurrentLoop.getName() << " (" << MinIterCount << ")\n");
+  auto *ExitBB = CurrentLoop.getExitingBlock();
+  if (!ExitBB)
+    return PreservedAnalyses::all();
+
+  // check if the branch instruction is the correct one
+  BranchInst *BI = dyn_cast<BranchInst>(ExitBB->getTerminator());
+  if (!BI || BI->isUnconditional())
+    return PreservedAnalyses::all();
+
+  ICmpInst *LoopCmpInstr = dyn_cast<ICmpInst>(BI->getCondition());
+  if (!LoopCmpInstr)
+    return PreservedAnalyses::all();
+
+  LLVM_DEBUG(dbgs() << "Branch Instruction Found: "; LoopCmpInstr->dump();
+             dbgs() << "\n");
+
+  if (LoopCmpInstr->getPredicate() == CmpInst::ICMP_EQ ||
+      LoopCmpInstr->getPredicate() == CmpInst::ICMP_NE) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountsToAssumptions-Warning: EQ and NE "
+                         "conditions are currently not supported!");
+    return PreservedAnalyses::all();
+  };
+
+  // get Scalar Evolution of the induction variable
+  unsigned IVOperand = -1;
+  const SCEVAddRecExpr *AddRecExpr =
+      getLoopCmpIVAddRecExpr(*LoopCmpInstr, SE, IVOperand);
+  if (!AddRecExpr) {
+    LLVM_DEBUG(
+        dbgs() << "LoopIterCountsToAssumptions-Warning: Could not extract "
+                  "SCEVAddRecExpr! Will not process Metadata\n");
+    return PreservedAnalyses::all();
+  }
+
+  LLVM_DEBUG(dbgs() << "Loop Metadata        : " << MinIterCount << "\n");
+
+  IRBuilder<> Builder(CurrentLoop.getHeader()->getTerminator());
+
+  SCEVExpander Expander(
+      *SE, CurrentLoop.getHeader()->getParent()->getParent()->getDataLayout(),
+      "expanded");
+
+  // get Minimum Iteration Value Expression
+  const SCEV *MinIterSCEV = AddRecExpr->evaluateAtIteration(
+      SE->getConstant(APInt(32, MinIterCount)), *SE);
+
+  // in the case of decrementing, add NSW flag, so that the expression can be
+  // simplified
+  if (auto *AE = dyn_cast<SCEVAddExpr>(MinIterSCEV)) {
+    SmallVector<const SCEV *> Ops(AE->operands());
+    MinIterSCEV = SE->getAddExpr(Ops, SCEV::FlagNSW);
+  }
+
+  Value *MinIterValue = Expander.expandCodeFor(
+      MinIterSCEV, LoopCmpInstr->getOperand(0)->getType(),
+      CurrentLoop.getHeader()->getTerminator());
+
+  LLVM_DEBUG(dbgs() << "MinIterValue: "; MinIterValue->dump());
+
+  Value *LoopAssertValue;
+  Value *LoopStartValue = Expander.expandCodeFor(
+      AddRecExpr->getStart(), LoopCmpInstr->getOperand(0)->getType(),
+      CurrentLoop.getHeader()->getTerminator());
+
+  // if the loop is decrementing, the Initial Start of the loop is already in
+  // the MinIterValue, therefore it does not have to be added again
+  if (needStartValue(*LoopStartValue, *MinIterValue)) {
+    LLVM_DEBUG(dbgs() << "Adding Loop Start Value: "; LoopStartValue->dump());
+    LoopAssertValue =
+        Builder.CreateAdd(LoopStartValue, MinIterValue, "", false, true);
+  } else {
+    LoopAssertValue = MinIterValue;
+  }
+
+  Value *Cmp;
+  if (IVOperand == 0) {
```

Comment:
We can vanish this if and have just one `CreateICmp` by the using a ternary operator around the operand index.

Reply:
or indeed `1 - IVOperand`

Reply:
Or define a variable once and for all for e.g. `IVValue` and `IVComparisonValue`

Reply:
I have changing positions of the operands depending on the IVOperand position, therefore I implemented it with ternary operators